# ==============================================================================
# REGRESSOR COMMIT DIFF
# ==============================================================================
# File: widget/gtk/WaylandSurface.h
# Commit: 9b286c5c32d1
# Full Hash: 9b286c5c32d1accc0025872508b34a1eeb0d2d64
# Author: stransky <stransky@redhat.com>
# Date: 2025-07-22 04:26:57
# Regressor Bug: 1978207
# File Overlap Count: 4
# Description:
#   Bug 1978207 [Wayland] Remove detached WaylandBuffer transactions from WaylandSurface r=emilio
#   
#   When WaylandBuffer transaction is detached from compositor, remove it from WaylandSurface.
#   
#   Differential Revision: https://phabricator.services.mozilla.com/D257976
# ==============================================================================

diff -r d73d4fbe1330 -r 9b286c5c32d1 widget/gtk/WaylandSurface.h
--- a/widget/gtk/WaylandSurface.h	Mon Jul 21 15:15:52 2025 +0000
+++ b/widget/gtk/WaylandSurface.h	Mon Jul 21 15:19:40 2025 +0000
@@ -128,12 +128,18 @@
   // on screen.
   bool AttachLocked(const WaylandSurfaceLock& aSurfaceLock,
                     RefPtr<WaylandBuffer> aBuffer);
+  bool IsBufferAttached(WaylandBuffer* aBuffer);
 
   // If there's any WaylandBuffer recently attached, detach it.
   // It makes the WaylandSurface invisible and it doesn't have any
   // content.
   void RemoveAttachedBufferLocked(const WaylandSurfaceLock& aProofOfLock);
 
+  // Remove deleted transaction from WaylandSurface, it may release
+  // referenced WaylandBuffer.
+  void RemoveTransactionLocked(const WaylandSurfaceLock& aSurfaceLock,
+                               RefPtr<BufferTransaction> aTransaction);
+
   // CommitLocked() is needed to call after some of *Locked() method
   // to submit the action to Wayland compositor by wl_surface_commit().
 
@@ -281,6 +287,8 @@
   void Commit(WaylandSurfaceLock* aProofOfLock, bool aForceCommit,
               bool aForceDisplayFlush);
 
+  BufferTransaction* GetNextTransactionLocked(
+      const WaylandSurfaceLock& aSurfaceLock, WaylandBuffer* aBuffer);
   // Force release/detele all transactions and wl_buffers attached to them.
   void ReleaseAllWaylandTransactionsLocked(WaylandSurfaceLock& aSurfaceLock);
 
@@ -350,6 +358,7 @@
   // there buffers live until compositor notify us that we
   // can release them.
   AutoTArray<RefPtr<BufferTransaction>, 3> mBufferTransactions;
+  uintptr_t mLatestAttachedBuffer = 0;
 
   // Indicates mSurface has buffer attached so we can attach subsurface
   // to it and expect to get frame callbacks from Wayland compositor.
